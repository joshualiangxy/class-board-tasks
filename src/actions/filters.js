export const sortByDeadline = () => ({ type: 'SORT_BY_DEADLINE' });

export const sortByName = () => ({ type: 'SORT_BY_NAME' });

export const setTextFilter = (text = '') => ({ type: 'SET_TEXT_FILTER', text });

export const toggleGrouped = () => ({ type: 'TOGGLE_GROUPED' });
