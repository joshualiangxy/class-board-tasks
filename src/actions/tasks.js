import firebase from '../firebase/firebase';

export const addPersonalTask = task => ({ type: 'ADD_PERSONAL_TASK', task });

export const startAddPersonalTask = task => {
  return (dispatch, getState) => {
    const uid = getState().auth.user.uid;
    const id = task.id;

    return firebase
      .firestore()
      .collection('users')
      .doc(uid)
      .collection('tasks')
      .doc(id)
      .set(task)
      .then(() => dispatch(addPersonalTask(task)));
  };
};

export const removePersonalTask = id => ({ type: 'REMOVE_PERSONAL_TASK', id });

export const startRemovePersonalTask = id => {
  return (dispatch, getState) => {
    const uid = getState().auth.user.uid;

    return firebase
      .firestore()
      .collection('users')
      .doc(uid)
      .collection('tasks')
      .doc(id)
      .delete()
      .then(() => dispatch(removePersonalTask(id)));
  };
};

export const editPersonalTask = (id, updates) => ({
  type: 'EDIT_PERSONAL_TASK',
  id,
  updates
});

export const startEditPersonalTask = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.user.uid;

    return firebase
      .firestore()
      .collection('users')
      .doc(uid)
      .collection('tasks')
      .doc(id)
      .set(updates)
      .then(dispatch(editPersonalTask(id, updates)));
  };
};

export const setTasks = tasks => ({ type: 'SET_TASK', tasks });

export const startSetTasks = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.user.uid;
    const tasks = [];
    const userRef = firebase.firestore().collection('users').doc(uid);
    const tasksRef = userRef.collection('tasks');

    return tasksRef
      .get()
      .then(querySnapshot =>
        querySnapshot.forEach(queryDocSnapshot =>
          tasks.push({ id: queryDocSnapshot.id, ...queryDocSnapshot.data() })
        )
      )
      .then(() => {
        userRef.get().then(userDocSnapshot => {
          const groupIds = userDocSnapshot.get('groups') || [];
          const groupsPromises = [];

          groupIds.forEach(gid => {
            const groupPromises = [];
            const groupRef = firebase.firestore().collection('groups').doc(gid);

            groupPromises.push(
              groupRef
                .get()
                .then(groupDocSnapshot => groupDocSnapshot.get('name'))
            );
            groupPromises.push(groupRef.collection('tasks').get());

            groupsPromises.push(
              Promise.all(groupPromises).then(
                ([groupName, queryTaskCollectionSnapshot]) =>
                  queryTaskCollectionSnapshot.forEach(queryTaskDocSnapshot =>
                    tasks.push({
                      groupName,
                      gid,
                      id: queryTaskDocSnapshot.id,
                      ...queryTaskDocSnapshot.data()
                    })
                  )
              )
            );

            return Promise.all(groupPromises);
          });
        });
      })
      .then(() => dispatch(setTasks(tasks)));
  };
};

export const toggleCompletedPersonal = (id, completedState) => ({
  type: 'TOGGLE_COMPLETED_PERSONAL',
  id,
  completedState
});

export const startToggleCompletedPersonal = (id, completedState) => {
  return (dispatch, getState) => {
    const uid = getState().auth.user.uid;

    return firebase
      .firestore()
      .collection('users')
      .doc(uid)
      .collection('tasks')
      .doc(id)
      .update({ completed: !completedState })
      .then(() => toggleCompletedPersonal(id, completedState));
  };
};

export const removeTaskData = () => ({ type: 'REMOVE_TASK_DATA' });
