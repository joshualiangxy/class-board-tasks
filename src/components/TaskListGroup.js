import React from 'react';
import TaskListItem from './TaskListItem';

const TaskListGroup = ({ group }) => {
  return (
    <div>
      <h3>{group.groupName}</h3>
      {group.tasks.map(task => (
        <TaskListItem key={task.id} task={task} showGroup={false} />
      ))}
    </div>
  );
};

export default TaskListGroup;
