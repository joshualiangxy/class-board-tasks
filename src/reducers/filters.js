const defaultState = { text: '', sortBy: 'deadline', grouped: true };

const filtersReducer = (state = defaultState, action) => {
  switch (action.type) {
    case 'SORT_BY_NAME':
      return { ...state, sortBy: 'name' };
    case 'SORT_BY_DEADLINE':
      return { ...state, sortBy: 'deadline' };
    case 'SET_TEXT_FILTER':
      return { ...state, text: action.text };
    case 'TOGGLE_GROUPED':
      return { ...state, grouped: !state.grouped };
    default:
      return state;
  }
};

export default filtersReducer;
