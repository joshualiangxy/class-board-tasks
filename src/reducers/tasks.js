const tasksReducer = (state = [], action) => {
  switch (action.type) {
    case 'ADD_TASK':
      return [...state, action.task];
    case 'REMOVE_TASK':
      return state.filter(task => task.id !== action.id);
    case 'EDIT_TASK':
      return state.map(task =>
        task.id === action.id
          ? {
              ...task,
              ...action.updates
            }
          : task
      );
    case 'SET_TASK':
      return action.tasks;
    default:
      return state;
  }
};

export default tasksReducer;
