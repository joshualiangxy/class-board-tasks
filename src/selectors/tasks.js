const getSortedTasks = (tasks, { text, sortBy, grouped }) => {
  const lowercaseText = text.toLowerCase();

  const processedTasks = tasks
    .filter(
      task =>
        task.title.toLowerCase().includes(lowercaseText) ||
        task.description.toLowerCase().includes(lowercaseText) ||
        task.module.toLowerCase().includes(lowercaseText)
    )
    .sort((taskOne, taskTwo) => {
      switch (sortBy) {
        case 'deadline':
          if (taskOne.deadline && taskTwo.deadline)
            return taskOne.deadline - taskTwo.deadline;
          else if (taskOne.deadline) return -1;
          else if (taskTwo.deadline) return 1;
          else return taskOne.title.localeCompare(taskTwo.title);
        case 'name':
          return taskOne.title.localeCompare(taskTwo.title);
      }
    });

  const sortedList = {};

  if (grouped) {
    processedTasks.forEach(task => {
      if (task.groupName) {
        const gid = task.gid;

        if (!sortedList[gid])
          sortedList[gid] = { groupName: task.groupName, tasks: [] };
        sortedList[gid].tasks.push(task);
      } else if (task.module) {
        const module = task.module;

        if (!sortedList[module])
          sortedList[module] = { groupName: task.module, tasks: [] };
        sortedList[module].tasks.push(task);
      } else {
        if (!sortedList.others)
          sortedList.others = { groupName: 'others', tasks: [] };
        sortedList.others.tasks.push(task);
      }
    });
  } else sortedList.others = { groupName: 'others', tasks: processedTasks };

  return sortedList;
};

export default getSortedTasks;
