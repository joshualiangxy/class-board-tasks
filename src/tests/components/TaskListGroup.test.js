import React from 'react';
import { shallow } from 'enzyme';
import TaskListGroup from '../../components/TaskListGroup';
import getSortedTasks from '../../selectors/tasks';
import tasks from '../fixtures/tasks';
import { filtersGrouped } from '../fixtures/filters';

let wrapper;
const sortedList = getSortedTasks(tasks, filtersGrouped);

beforeEach(() => {
  wrapper = shallow(
    <TaskListGroup group={sortedList[Object.keys(sortedList)[0]]} />
  );
});

describe('render', () => {
  it('should render TaskListGroup', () => expect(wrapper).toMatchSnapshot());

  it('should render TaskListGroup with alt group', () => {
    wrapper.setProps({ group: sortedList[Object.keys(sortedList)[1]] });

    expect(wrapper).toMatchSnapshot();
  });
});
