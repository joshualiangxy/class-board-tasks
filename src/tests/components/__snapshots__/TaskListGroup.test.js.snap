// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render should render TaskListGroup 1`] = `
<div>
  <h3>
    CS2030
  </h3>
  <Connect(TaskListItem)
    dashboard={true}
    key="5feb18c0-78a8-49b6-8584-75b8c1b40abc"
    showGroup={false}
    task={
      Object {
        "completed": false,
        "deadline": 1591329600000,
        "description": "Catch up on lectures!",
        "id": "5feb18c0-78a8-49b6-8584-75b8c1b40abc",
        "module": "CS2030",
        "title": "Watch lecture",
      }
    }
  />
  <Connect(TaskListItem)
    dashboard={true}
    key="a1050220-fa88-4dc2-aa9d-9521ae95de47"
    showGroup={false}
    task={
      Object {
        "completed": false,
        "description": "",
        "id": "a1050220-fa88-4dc2-aa9d-9521ae95de47",
        "module": "CS2030",
        "title": "Do lab",
      }
    }
  />
</div>
`;

exports[`render should render TaskListGroup with alt group 1`] = `
<div>
  <h3>
    others
  </h3>
  <Connect(TaskListItem)
    dashboard={true}
    key="d2b10492-853d-4290-baf8-cb47cd4ec04e"
    showGroup={false}
    task={
      Object {
        "completed": true,
        "description": "",
        "id": "d2b10492-853d-4290-baf8-cb47cd4ec04e",
        "module": "",
        "title": "Read a book everyday",
      }
    }
  />
</div>
`;
